wp-content/plugins/tapestry/controller/class.user-controller.php

```diff
@@ -0,0 +1,105 @@
+     * @param type @userId the user's id
+     * @param type @postId the post's ID
+     * @param type @progressData json data for progress of each node
+    */
+    public function updateProgress($userId, $postId, $progressData) {
+        $this->checkUserAndPostId($userId, $postId);  
+
+        if ($this->isJson($progressData)) {
+            $progressData = json_decode($progressData);
+        } else {
+            throw new Exception('Invalid json');
+        }
+
+        $this->updateUserProgress($userId, $postId, $progressData);
+    }
+
+    /** 
+     * Get User's video progress for a tapestry post
+     * @param type @userId the user's id
+     * @param type @postId the post's ID
+    */
+    public function getProgress($userId, $postId) {
+        $this->checkUserAndPostId($userId, $postId);  
+        return $this->getUserProgress($userId, $postId);
+    }
+
+    /** 
+     * Update User's h5p video setting for a tapestry post
+     * @param type @userId the user's id
+     * @param type @postId the post's ID
+    */
+    public function updateH5PSetting($userId, $postId, $h5pSettingsData) {
+        $this->checkUserAndPostId($userId, $postId);  
+
+        if ($this->isJson($h5pSettingsData)) {
+            $h5pSettingsData = json_decode($h5pSettingsData);
+        } else {
+            throw new Exception('Invalid json');
+        }
+
+        $this->updateUserH5PSetting($userId, $postId, $h5pSettingsData);
+    }
+
+    /** 
+     * Get User's h5p video setting for a tapestry post
+     * @param type @userId the user's id
+     * @param type @postId the post's ID
+    */
+    public function getH5PSetting($userId, $postId) {
+        $this->checkUserAndPostId($userId, $postId);  
+        return $this->getUserH5PSetting($userId, $postId);
+    }
+
+    private function updateUserProgress($userId, $postId, $progressData) {
+        update_user_meta($userId, 'progress_data_' . $postId, $progressData);
+    }
+
+    private function getUserProgress($userId, $postId) {
+        $progress = get_user_meta($userId, 'progress_data_' . $postId, true);
+        return $progress;
+    }
+
+    private function updateUserH5PSetting($userId, $postId, $h5pSettingsData) {
+        update_user_meta($userId, 'h5p_setting_' . $postId, $h5pSettingsData);
+    }
+
+    private function getUserH5PSetting($userId, $postId) {
+        $settings = get_user_meta($userId, 'h5p_setting_' . $postId, true);
+        return $settings;
+    }
+
+    // Helpers
+    private function checkUserAndPostId( $userId, $postId) {
+        if (!isset($userId)) {
+            throw new Exception('userId is invalid');
+        }
+
+        // Not the current user. Comment out when testing the endpoints on Postman
+        if ($userId != apply_filters('determine_current_user', false)) {
+            throw new Exception('userId is invalid');
+        }
+
+        if (!isset($postId)) {
+            throw new Exception('postId is invalid');
+        }
```

Comment:
1. Added a post type check.
2.  Currently this function is used for both UPDATE and GET. For the UPDATE situation, if postId does not exist in db, it is valid because it'll create the post ID. So I added a new function (postsExist) to check if postId exists in the db. I'll then use this function for the GET situations.